### ISSUE TITLE
When setting the validation set be the same with the training set avg_iou != avg_iou_val

### ISSUE THREAD

**Issue by CharlieLeee** [loicland/superpoint_graph#186]
I set the training dataset the same with the validation dataset in order to find out the best model that fits the training data, however, when monitoring the training process, I found out the validation ious are quite different from training ious, is that normal?

**Comment by loicland** [comment]
It's normal that they should be a little different (the method is stochastic), but not very different.

- is the difference always in the same direction?
- which dataset are you using, a custom one?
- check in debug mode the clouds[1] values in the training loop as well ad embeddings outputs by the pointnet inside the evaluation loop.
- set `--test_multisamp_n` to 1 and see if the problem persists

Let me know!

**Comment by CharlieLeee** [comment]
At certain epoch, they are in the opposite direction
<img width="1153" alt="train val 2019-12-17 02_41_01" src="https://user-images.githubusercontent.com/33972674/70966627-d0ffd380-20ce-11ea-9404-c4b6fbec73b9.png">

The dataset was split from the semantic3d dataset and I used the command:
```
--use_val_set 
```

I'll try to set `--test_multisamp_n` to 1 and see the result

Thanks

**Comment by loicland** [comment]
I think I have missunderstood what you did. What is your training and valdiation set comprised of?

The figure you posted seems ok to me. The method is doubly stochastic: random subgraph are extracted during training, so it is seen differently in train and eval mode; the superpoints are randomly sampled.

**Comment by CharlieLeee** [comment]
@loicland Thanks for the explanation! the training set and validation set are exactly the same, so I expected the same value for val iou and train iou, so as you've said, these two processes were done on different downsamples of the original data? So I should expect different results?

**Comment by loicland** [comment]
Different downsampling for superpoints, but also at training time only a subgraph of the full spg is extracted to allows for larger batch and graph augmentation. As you can see however, even if train and eval are not exactly identical they evolves similarily.

**Comment by CharlieLeee** [comment]
@loicland Right, thanks! also, another question, if I want to use validation set and save the best model during the validation, should I alter the argument "-save_nth_epoch" to be larger than the epoch I'm training, are there any other parameters/code I should change if I want to use a validation set?

**Comment by CharlieLeee** [comment]
@loicland Hi loicland, would you please tell me how to train the model using validation set, and test with the best performance model on the validation set, it seems that in the 'main.py', the program will automatically save the model at the last epoch as the final model?

**Comment by loicland** [comment]
Hi,

use `--use_val_set 1` to save the epoch that performs best on the validation set.

**Comment by CharlieLeee** [comment]
@loicland should I also add --use_val_set 1 when performing the testing?

**Comment by loicland** [comment]
No need because you set epochs to -1 for testing. Note that when doping the inference on semantic3d  the metrics will be nonsensical because the ground truth is withheld. you should use the function `write_semantic3d` in `\partition` to write the labels files and upload them on the semantic3d servers.

**Comment by CharlieLeee** [comment]
![image](https://user-images.githubusercontent.com/33972674/71523730-36d93000-2905-11ea-855a-807814db3651.png)

@loicland After training with validation and testing at the end of the training, the model's epoch changed to 500 which is the number of training epoch, does that mean the training save only the model from the last epoch?
Also, as for the "epoch" in the scores*.json file
![image](https://user-images.githubusercontent.com/33972674/71523761-5bcda300-2905-11ea-80e0-ffc1db4b61e4.png)
does this imply the total number of epoch or the best epoch?
